nn/conformer.py

```diff
@@ -0,0 +1,205 @@
+"""
+Conformer code.
+Ref: https://arxiv.org/abs/2005.08100
+"""
+
+from typing import Tuple, List, Union
+
+from . import Module, ModuleList, LayerRef, Linear, dropout, layer_norm, batch_norm, Conv, swish, glu, split_dims, \
+  merge_dims, pool
+
+
+class _PositionwiseFeedForward(Module):
+  """
+  Conformer position-wise feedforward neural network layer
+      FF -> Activation -> Dropout -> FF
+  """
+
+  def __init__(self, d_model: int, d_ff: int, dropout: float, activation, l2: float = 0.0):
+    """
+    :param d_model:
+    :param d_ff:
+    :param dropout:
+    :param activation:
+    :param l2:
+    """
+    super().__init__()
+
+    self.dropout = dropout
+    self.activation = activation
+
+    self.linear1 = Linear(n_out=d_ff, l2=l2)
+    self.linear2 = Linear(n_out=d_model, l2=l2)
+
+  def forward(self, inp: LayerRef) -> LayerRef:
+    return self.linear2(dropout(self.activation(self.linear1(inp)), dropout=self.dropout))
+
+
+class _ConformerConvBlock(Module):
+  """
+  Conformer convolution block
+      FF -> GLU -> depthwise conv -> BN -> Swish -> FF
+  """
+
+  def __init__(self, d_model: int, kernel_size: Tuple[int], l2: float = 0.0):
+    """
+    :param d_model:
+    :param kernel_size:
+    :param l2:
+    """
+    super().__init__()
+
+    self.positionwise_conv1 = Linear(n_out=d_model * 2, l2=l2)
```

Comment:
I wonder, as you do GLU, whether it is maybe easier to use:
```suggestion
    self.positionwise_conv1a = Linear(n_out=d_model, l2=l2)
    self.positionwise_conv1b = Linear(n_out=d_model, l2=l2)
```